routers/web/web.go

```diff
@@ -1078,4 +1078,13 @@ func RegisterRoutes(m *web.Route) {
 	if setting.API.EnableSwagger {
 		m.Get("/swagger.v1.json", SwaggerV1Json)
 	}
+	m.NotFound(func(w http.ResponseWriter, req *http.Request) {
+		escapedPath := req.URL.EscapedPath()
+		if len(escapedPath) > 1 && escapedPath[len(escapedPath)-1] == '/' {
+			http.Redirect(w, req, setting.AppSubURL+escapedPath[:len(escapedPath)-1], http.StatusFound)
```

Comment:
```suggestion
			http.Redirect(w, req, setting.AppSubURL+escapedPath[:len(escapedPath)-1], http.StatusPermanentRedirect)
```

Reply:
All of our other Redirects are StatusFound

Reply:
Would using `escapedPath` cause problems for URLs like `/a%20b` ?

Reply:
We may need more testcases.

Reply:
And we need a return here.

Reply:
We forgot the `return` too many times. I would propose to make the framework like this:

```
func Handler(...) (Responser) {
    return ctx.NotFound(...)
}
```

Reply:
> Would using `escapedPath` cause problems for URLs like `/a%20b` ?

Why would escape path cause problems? It's precisely what you should use in that case!

Reply:
> 

Return has been added.

Reply:
> > Would using `escapedPath` cause problems for URLs like `/a%20b` ?
> 
> Why would escape path cause problems? It's precisely what you should use in that case!

No, I do not think it is the correct case. The URL in `Location` header should be encoded correctly.

```go
func main() {
	http.HandleFunc("/", func (w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, "/a b", 302)
	})
	_ = http.ListenAndServe(":3000", nil)
}
```

![image](https://user-images.githubusercontent.com/2114189/147049073-2426be74-86bb-408a-b8b9-acddabc077da.png)

Reply:
Oh, you are right. The escaped path doesn't decode other encodings.

Reply:
> All of our other Redirects are StatusFound

That does not make this one right.

MDN:

> Even if the specification requires the method (and the body) not to be altered when the redirection is performed, not all user-agents conform here - you can still find this type of bugged software out there. It is therefore recommended to set the 302 code only as a response for GET or HEAD methods and to use 307 Temporary Redirect instead, as the method change is explicitly prohibited in that case.

A POST to `/foo/` may become a GET `/foo` with the current code.

Reply:
Good catch @KN4CK3R

Reply:
PR updated